Hash long type columns

Hi, 
please, have you tried to hash long type columns? If you have column definition in JDBC input plugin like this

```yaml
in:
   type: db2
   driver_path: ../airflow/extra/db2jcc4.jar
   host: db2
   port: 50001
   user: 
   password: 
   database: G;
   fetch_rows: 10000
   table: CREDIT_ACCOUNT
   default_timezone: CET
   select: CREDIT_ACCOUNT_ID as CREDIT_ACCOUNT_ID,NAME as NAME,COMMENT as COMMENT,COMPANY_G2ID as COMPANY_G2ID
   column_options:
      CREDIT_ACCOUNT_ID: {value_type: string}
      NAME: {value_type: string}
      COMMENT: {value_type: string}
      COMPANY_G2ID: {value_type: long}
```

I try to define the hash filter this way and hash column COMPANY_G2ID as long type

```yaml
filters:
  - type: hash
    columns:
      - {name: NAME, algorithm: SHA-256}
      - {name: COMMENT, algorithm: SHA-256}
      - {name: COMPANY_G2ID, algorithm: SHA-256}
```

but I always get 

```
Caused by: java.lang.IllegalStateException: Not reach here
	at org.embulk.spi.PageBuilder$AbstractColumnValue.setLong(PageBuilder.java:348)
	at org.embulk.spi.PageBuilder$Row.setLong(PageBuilder.java:291)
	at org.embulk.spi.PageBuilder$Row.access$300(PageBuilder.java:248)
	at org.embulk.spi.PageBuilder.setLong(PageBuilder.java:83)
	at org.embulk.spi.PageBuilder.setLong(PageBuilder.java:79)
	at org.embulk.filter.hash.HashFilterPlugin$open$1.setValue(HashFilterPlugin.kt:93)
	at org.embulk.filter.hash.HashFilterPlugin$open$1.add(HashFilterPlugin.kt:69)
	at org.embulk.exec.LocalExecutorPlugin$ScatterTransactionalPageOutput$OutputWorker.call(LocalExecutorPlugin.java:353)
	at org.embulk.exec.LocalExecutorPlugin$ScatterTransactionalPageOutput$OutputWorker.call(LocalExecutorPlugin.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.lang.IllegalStateException: Not reach here
		... 13 more

Error: java.lang.IllegalStateException: Not reach here
```

Any ideas, how to solve this? Everything works fine, if I change the type of COMPANY_G2ID to string. 

Thank you very much!

--- a/build.gradle
+++ b/build.gradle
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlinVersion = '1.1.1'
+    ext.kotlinVersion = '1.2.31'
     repositories {
         mavenCentral()
         jcenter()

--- a/build.gradle
+++ b/build.gradle
@@ -22,16 +22,16 @@ repositories {
     maven { url 'http://kamatama41.github.com/maven-repository/repository' }
 }
 
-sourceCompatibility = 1.7
-targetCompatibility = 1.7
+sourceCompatibility = 1.8
+targetCompatibility = 1.8
 
 dependencies {
     compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlinVersion"
-    testCompile "com.github.kamatama41:embulk-test-helpers:0.3.2"
+    testCompile "com.github.kamatama41:embulk-test-helpers:0.4.0"
 }
 
 embulk {
-    version = "0.8.18"
+    version = "0.9.7"
     category = "filter"
     name = "hash"
     authors = ["Shinichi Ishimura"]

--- a/gradle.properties
+++ b/gradle.properties
@@ -1,1 +1,1 @@
-version=0.3.3-SNAPSHOT
+version=0.4.0-SNAPSHOT

--- a/src/main/kotlin/org/embulk/filter/hash/HashFilterPlugin.kt
+++ b/src/main/kotlin/org/embulk/filter/hash/HashFilterPlugin.kt
@@ -66,12 +66,12 @@ class HashFilterPlugin : FilterPlugin {
             override fun add(page: Page) {
                 reader.setPage(page)
                 while (reader.nextRecord()) {
-                    setValue()
+                    setRowValue()
                     builder.addRecord()
                 }
             }
 
-            private fun setValue() {
+            private fun setRowValue() {
                 for (inputColumn in inputSchema.columns) {
                     if (reader.isNull(inputColumn)) {
                         builder.setNull(inputColumn)

--- a/src/main/kotlin/org/embulk/filter/hash/HashFilterPlugin.kt
+++ b/src/main/kotlin/org/embulk/filter/hash/HashFilterPlugin.kt
@@ -78,37 +78,47 @@ class HashFilterPlugin : FilterPlugin {
                         continue
                     }
 
-                    // Write the original data
-                    val inputValue: Any = when (inputColumn.type) {
+                    when (inputColumn.type) {
                         Types.STRING -> {
-                            reader.getString(inputColumn).apply { builder.setString(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getString, builder::setString)
                         }
                         Types.BOOLEAN -> {
-                            reader.getBoolean(inputColumn).apply { builder.setBoolean(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getBoolean, builder::setBoolean)
                         }
                         Types.DOUBLE -> {
-                            reader.getDouble(inputColumn).apply { builder.setDouble(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getDouble, builder::setDouble)
                         }
                         Types.LONG -> {
-                            reader.getLong(inputColumn).apply { builder.setLong(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getLong, builder::setLong)
                         }
                         Types.TIMESTAMP -> {
-                            reader.getTimestamp(inputColumn).apply { builder.setTimestamp(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getTimestamp, builder::setTimestamp)
                         }
                         Types.JSON -> {
-                            reader.getJson(inputColumn).apply { builder.setJson(inputColumn, this) }
+                            setColumnValue(inputColumn, reader::getJson, builder::setJson)
                         }
                         else -> {
                             throw DataException("Unexpected type:" + inputColumn.type)
                         }
                     }
+                }
+            }
 
-                    // Overwrite the column if it's hash column.
-                    hashColumnMap[inputColumn.name]?.let { hashColumn ->
-                        val outputColumn = outputColumnMap[hashColumn.newName.or(inputColumn.name)]
-                        val hashedValue = generateHash(inputValue.toString(), hashColumn.algorithm.get())
-                        builder.setString(outputColumn, hashedValue)
-                    }
+            private fun <T> setColumnValue(
+                    inputColumn: Column,
+                    getter: (inputColumn: Column) -> T,
+                    setter: (inputColumn: Column, value: T) -> Unit
+            ) {
+                val inputValue = getter(inputColumn)
+
+                hashColumnMap[inputColumn.name]?.let { hashColumn ->
+                    // Write hashed value if it's hash column.
+                    val outputColumn = outputColumnMap[hashColumn.newName.or(inputColumn.name)]
+                    val hashedValue = generateHash(inputValue.toString(), hashColumn.algorithm.get())
+                    builder.setString(outputColumn, hashedValue)
+                } ?: run {
+                    // Write the original data
+                    setter(inputColumn, inputValue)
                 }
             }
 

